core/deployment/src/main/java/io/quarkus/deployment/recording/BytecodeRecorderImpl.java

```diff
@@ -1445,24 +1446,34 @@ ResultHandle createValue(MethodContext context, MethodCreator method, ResultHand
                             Arrays.stream(finalCtorHandles).map(m -> context.loadDeferred(m))
                                     .toArray(ResultHandle[]::new));
                 } else {
-                    try {
-                        param.getClass().getDeclaredConstructor();
-                        out = method.newInstance(ofConstructor(param.getClass()));
-                    } catch (NoSuchMethodException e) {
-                        //fallback for collection types, such as unmodifiableMap
-                        if (SortedMap.class.isAssignableFrom(expectedType)) {
-                            out = method.newInstance(ofConstructor(TreeMap.class));
-                        } else if (Map.class.isAssignableFrom(expectedType)) {
-                            out = method.newInstance(ofConstructor(LinkedHashMap.class));
-                        } else if (List.class.isAssignableFrom(expectedType)) {
-                            out = method.newInstance(ofConstructor(ArrayList.class));
-                        } else if (SortedSet.class.isAssignableFrom(expectedType)) {
-                            out = method.newInstance(ofConstructor(TreeSet.class));
-                        } else if (Set.class.isAssignableFrom(expectedType)) {
-                            out = method.newInstance(ofConstructor(LinkedHashSet.class));
+                    if (List.class.isAssignableFrom(param.getClass())) {
+                        // list is a common special case, so let's handle it
+                        List listParam = (List) param;
+                        if (listParam.isEmpty()) {
+                            out = method.invokeStaticMethod(COLLECTIONS_EMPTY_LIST); // TODO: Should we do this? There is a risk of breaking recorders that don't treat the params as immutable
```

Comment:
This is the one I am not really sure about. It saves allocations, but it might too unsafe

Reply:
IMHO this should always be immutable, or always mutable, not sometimes mutable except when it is empty.

I would just leave it as an empty ArrayList for now, immutable would be better but it might break universe extensions and it does not really matter.

Reply:
Fixed